### PR TITLE
Various builder bugs

### PR DESCRIPTION
**Pull request by pandafulmanda**
### Fixes include

- [x] "Set as default" will show as expected for valid times
- [x] "Default set" message will show for a bit after default set
- [x] cursor does not jump when typing time
- [x] default times fill in as default times as expected when making new assignment and switching between all vs individual periods
- [x] show individual periods by default if default times for periods are different from each other
- [x] only subscribe to job status when publish is in progress
- [x] cancel will better track if there's been changes
- [x] if new assignment being created without default due date, due date will remain blank until deliberately selected
- [x] datepicker blur will return input label to expected location

### CODE REVIEW

File: src/components/task-plan/builder/date-time.cjsx [hunk 1]
```diff
@@ -41,30 +48,58 @@ DateTime = React.createClass
...
+    prevState.isSetting and
+      not @state.isSetting and
+      not prevProps.isTimeDefault and
+      @props.isTimeDefault
+
+  isDateValid: ->
+    @state?.date? and _.isEmpty(@refs?.date?.state?.errors)
+
+  isTimeValid: ->
+    @state?.time? and _.isEmpty @refs?.time?.refs?.timeInput?.state?.errors
```
Comment (philschatz) [inline src/components/task-plan/builder/date-time.cjsx]: Parentheses missing ; (

File: src/components/task-plan/builder/date-time.cjsx [hunk 2]
```diff
@@ -41,30 +48,58 @@ DateTime = React.createClass
...
+
+  isTimeDefault: ->
+    return true if _.isUndefined(@state?.time)
+    {defaultTime} = @props
+    TimeHelper.makeMoment(@state.time, 'HH:mm').isSame(TimeHelper.makeMoment(defaultTime, 'HH:mm'), 'minute')
```
Comment (philschatz) [inline src/components/task-plan/builder/date-time.cjsx]: Don't we store the time formats in a centralized place?

File: src/components/task-plan/builder/index.cjsx [hunk 3]
```diff
@@ -146,42 +150,46 @@ TaskPlanBuilder = React.createClass
...
-    taskingDueAt = TaskPlanStore.getDueAt(@props.id) or @getQueriedDueAt()
-
-    if taskingDueAt
-      @setDueAt(taskingDueAt)
+    if savedAllTaskings
+      TaskPlanActions.replaceTaskings(@props.id, savedAllTaskings)
     else
-      TaskPlanActions.clearDueAt(@props.id)
+      periods = _.map CourseStore.getPeriods(@props.courseId), (period) -> id: period.id
+      TaskPlanActions.setDefaultTimesForCourse(@props.id, courseId, periods)
```
Comment (nathanstitt) [inline src/components/task-plan/builder/index.cjsx]: nitpick: could use `_.pluck CourseStore.getPeriods(@props.courseId), 'id'` vs map

File: src/components/task-plan/builder/index.cjsx [hunk 4]
```diff
@@ -88,27 +93,31 @@ TaskPlanBuilder = React.createClass
...
+    nextState.showingPeriods = not (commonDates and hasAllTaskings) unless isNewPlan
 
     # Inform the store of the available periods
-    TaskPlanActions.setPeriods(planId, courseId, periods)
+    TaskPlanActions.setPeriods(id, courseId, periods,
+      not isInitialized, not (nextState.showingPeriods or showingPeriods))
 
-    unless isNewPlan
-      @setState({showingPeriods: not (commonDates and hasAllTaskings)})
+    @setState(nextState) unless _.isEmpty(nextState)
```
Comment (philschatz) [inline src/components/task-plan/builder/index.cjsx]: oh, nice use of `isEmpty`!

File: src/components/tutor-input.cjsx [hunk 5]
```diff
@@ -192,7 +198,7 @@ TutorDateInput = React.createClass
       displayValue = value.format(TutorDateFormat)
 
     <div className={wrapperClasses}>
-      <input type='text' disabled readonly className={classes} value={displayValue}/>
+      <input type='text' disabled readOnly className={classes} value={displayValue}/>
       <div className="floating-label">{@props.label}</div>
```
Comment (nathanstitt) [inline src/components/tutor-input.cjsx]: nice catch!  I always forget to camel case the react props

File: src/components/task-plan/builder/index.cjsx [hunk 6]
```diff
@@ -146,42 +150,46 @@ TaskPlanBuilder = React.createClass
 
   setAllPeriods: ->
     {courseId} = @props
+    {showingPeriods, savedAllTaskings} = @state
 
     #save current taskings
-    if @state.showingPeriods
```
Comment (philschatz) [inline src/components/task-plan/builder/index.cjsx]: Woah, does this code do the same thing as before? /cc @jivewise

File: src/flux/task-plan.coffee [hunk 7]
```diff
@@ -122,38 +122,66 @@ TaskPlanConfig =
...
+      tasking ?= target_id: period.id, target_type:'period'
+
+      opens_at = period.opens_at or tasking.opens_at
+      due_at = period.due_at or tasking.due_at
+
+      period.opens_at = TimeHelper.makeMoment(opens_at).format(TimeHelper.ISO_DATE_FORMAT) if opens_at?
+      period.due_at = TimeHelper.makeMoment(due_at).format(TimeHelper.ISO_DATE_FORMAT) if due_at?
+
+      periodTimes = @setDefaultTimes(course, period, useCourseDefault) if opens_at? or due_at?
+
```
Comment (nathanstitt) [inline src/flux/task-plan.coffee]: idea for future cleanup:  add a `TimeHelper.toIsoDateString` (or something better named) that you can pass a `string` or `date` to and get back ISO_DATE_FORMAT formatted string.

**Comment by nathanstitt** [comment]
Looks like it's a good improvement @pandafulmanda !

File: src/components/task-plan/builder/index.cjsx (hunk 6 above)
Comment (pandafulmanda) [inline src/components/task-plan/builder/index.cjsx]: yes, it wasnt really doing anything when it was in there, `@state.showingPeriods` was always true

File: src/components/tutor-input.cjsx [hunk 8]
```diff
@@ -31,6 +31,9 @@ TutorInput = React.createClass
     errors = @props.validate(@props.default)
     errors: errors or []
 
+  componentDidUpdate: (prevProps, prevState) ->
+    @props.onUpdated?(@state) unless _.isEqual(prevState, @state)
+
```
Comment (pandafulmanda) [inline src/components/tutor-input.cjsx]: oh, yah :( good one! teehehehe i tend to neglect the `propTypes` and i've notice you always do a good job with them

File: src/components/task-plan/builder/index.cjsx (hunk 3 above)
Comment (pandafulmanda) [inline src/components/task-plan/builder/index.cjsx]: oh yes, i would prefer that as well, not sure why this is `map`. great catch!

File: src/components/task-plan/builder/date-time.cjsx (hunk 2 above)
Comment (pandafulmanda) [inline src/components/task-plan/builder/date-time.cjsx]: oh hmm, yah i could move this into `TimeHelper` like the others

File: src/components/task-plan/builder/date-time.cjsx (hunk 1 above)
Comment (pandafulmanda) [inline src/components/task-plan/builder/date-time.cjsx]: oops, right next to each other too!

**Pull request merged**